File: src/Analyzer.hs
module Analyzer
    ( analyzeDirectory
    , analyzeWithConfig
    ) where

import Control.Monad (forM)
import System.FilePath (normalise)
import qualified Data.Text as Text
import Config (normalizeFormats)
import Metrics (analyzeFile)
import Report (buildReport, writeReports)
import Scanner (scanForSources)
import Types (Config (..), OutputFormat (..))

analyzeDirectory :: FilePath -> IO [FilePath]
analyzeDirectory dir = analyzeWithConfig Config
    { configRoot = dir
    , configOutputDir = "reports"
    , configFormats = normalizeFormats [OutputFormat (Text.pack "text"), OutputFormat (Text.pack "json")]
    , configFollowSymlinks = False
    }

analyzeWithConfig :: Config -> IO [FilePath]
analyzeWithConfig config = do
    let root = normalise (configRoot config)
    sources <- scanForSources (configFollowSymlinks config) root
    metrics <- forM sources analyzeFile
    report <- buildReport root metrics
    writeReports (configOutputDir config) (configFormats config) report

File: haskmaintain/analyzer.py
"""Core analysis pipeline."""

from __future__ import annotations

from pathlib import Path
from typing import List

from haskmaintain.config import Config, normalize_formats
from haskmaintain.metrics import FileMetrics, analyze_file
from haskmaintain.report import build_report, write_reports
from haskmaintain.scanner import scan_sources


def analyze(config: Config) -> List[Path]:
    root = config.root
    formats = normalize_formats(config.formats)
    sources = scan_sources(root, follow_symlinks=config.follow_symlinks)
    metrics: List[FileMetrics] = [analyze_file(path) for path in sources]
    report = build_report(root, metrics)
    return write_reports(config.output_dir, formats, report)
